diff --git a/pipenv/vendor/pipdeptree.py b/pipenv/vendor/pipdeptree.py
index 6f6f2a10..cd67538c 100644
--- a/pipenv/vendor/pipdeptree.py
+++ b/pipenv/vendor/pipdeptree.py
@@ -13,13 +13,11 @@ try:
 except ImportError:
     from ordereddict import OrderedDict
 
-try:
-    from pipenv.patched.notpip._internal.utils.misc import get_installed_distributions
-    from pipenv.patched.notpip._internal.operations.freeze import FrozenRequirement
-except ImportError:
-    from pipenv.patched.notpip import get_installed_distributions, FrozenRequirement
+pardir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+sys.path.append(pardir)
+from pipenv.vendor.pip_shims import get_installed_distributions, FrozenRequirement
 
-from pipenv.patched.notpip._vendor import pkg_resources
+import pkg_resources
 # inline:
 # from graphviz import backend, Digraph
 
