diff --git a/pipenv/vendor/pipdeptree.py b/pipenv/vendor/pipdeptree.py
index 6e2c3b3..c392e47 100644
--- a/pipenv/vendor/pipdeptree.py
+++ b/pipenv/vendor/pipdeptree.py
@@ -20,6 +20,8 @@ try:
 except ImportError:
     from collections import Mapping

+pardir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+sys.path.append(pardir)
 from pip._vendor import pkg_resources
 try:
     from pip._internal.operations.freeze import FrozenRequirement
