diff --git a/pipenv/vendor/pipdeptree.py b/pipenv/vendor/pipdeptree.py
index 6e2c3b3..4b92b02 100644
--- a/pipenv/vendor/pipdeptree.py
+++ b/pipenv/vendor/pipdeptree.py
@@ -20,7 +20,9 @@ try:
 except ImportError:
     from collections import Mapping

+pardir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+sys.path.append(pardir)
from pipenv.patched.pip._vendor import pkg_resources
 try:
     from pipenv.patched.pip._internal.operations.freeze import FrozenRequirement
 except ImportError:
