diff --git a/pipenv/vendor/tomlkit/container.py b/pipenv/vendor/tomlkit/container.py
index 37014921..a0014c6e 100644
--- a/pipenv/vendor/tomlkit/container.py
+++ b/pipenv/vendor/tomlkit/container.py
@@ -189,7 +189,12 @@ class Container(dict):
         if idx is None:
             raise NonExistentKey(key)
 
-        self._body[idx] = (None, Null())
+        old_data = self._body[idx][1]
+        trivia = getattr(old_data, "trivia", None)
+        if trivia and getattr(trivia, "comment", None):
+            self._body[idx] = (None, trivia)
+        else:
+            self._body[idx] = (None, Null())
 
         super(Container, self).__delitem__(key.key)
 
