diff --git a/pipenv/vendor/tomlkit/items.py b/pipenv/vendor/tomlkit/items.py
index 6617c522..9e746a8c 100644
--- a/pipenv/vendor/tomlkit/items.py
+++ b/pipenv/vendor/tomlkit/items.py
@@ -50,7 +50,7 @@ def item(value, _parent=None, _sort_keys=False):
             val[k] = item(v, _parent=val, _sort_keys=_sort_keys)
 
         return val
-    elif isinstance(value, list):
+    elif isinstance(value, (tuple, list)):
         if value and isinstance(value[0], dict):
             a = AoT([])
         else:
