diff --git a/pipenv/vendor/vistir/spin.py b/pipenv/vendor/vistir/spin.py
index 2a848922..57a90277 100644
--- a/pipenv/vendor/vistir/spin.py
+++ b/pipenv/vendor/vistir/spin.py
@@ -5,6 +5,7 @@ import os
 import signal
 import sys
 
+import colorama
 import cursor
 import six
 
@@ -31,6 +32,7 @@ CLEAR_LINE = chr(27) + "[K"
 
 class DummySpinner(object):
     def __init__(self, text="", **kwargs):
+        colorama.init()
         self.text = to_native_string(text)
         self.stdout = kwargs.get("stdout", sys.stdout)
         self.stderr = kwargs.get("stderr", sys.stderr)
@@ -112,7 +114,6 @@ class VistirSpinner(base_obj):
         """
 
         self.handler = handler
-        import colorama
         colorama.init()
         sigmap = {}
         if handler:
