diff --git a/pipenv/patched/notpip/_internal/pep425tags.py b/pipenv/patched/notpip/_internal/pep425tags.py
index 3c760ca3..3b11b965 100644
--- a/pipenv/patched/notpip/_internal/pep425tags.py
+++ b/pipenv/patched/notpip/_internal/pep425tags.py
@@ -178,7 +178,7 @@ def is_manylinux1_compatible():
         pass
 
     # Check glibc version. CentOS 5 uses glibc 2.5.
-    return pip._internal.utils.glibc.have_compatible_glibc(2, 5)
+    return pipenv.patched.notpip._internal.utils.glibc.have_compatible_glibc(2, 5)
 
 
 def is_manylinux2010_compatible():
@@ -196,7 +196,7 @@ def is_manylinux2010_compatible():
         pass
 
     # Check glibc version. CentOS 6 uses glibc 2.12.
-    return pip._internal.utils.glibc.have_compatible_glibc(2, 12)
+    return pipenv.patched.notpip._internal.utils.glibc.have_compatible_glibc(2, 12)
 
 
 def get_darwin_arches(major, minor, machine):
