diff --git a/pipenv/patched/pip/__main__.py b/pipenv/patched/pip/__main__.py
index 24d42002b..b424ab203 100644
--- a/pipenv/patched/pip/__main__.py
+++ b/pipenv/patched/pip/__main__.py
@@ -26,6 +26,13 @@ if __name__ == "__main__":
     warnings.filterwarnings(
         "ignore", category=DeprecationWarning, module=".*packaging\\.version"
     )
+    import importlib.util
+    import sys
+    spec = importlib.util.spec_from_file_location(
+        "pipenv", location=os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "__init__.py"))
+    pipenv = importlib.util.module_from_spec(spec)
+    sys.modules["pipenv"] = pipenv
+    spec.loader.exec_module(pipenv)
     from pipenv.patched.pip._internal.cli.main import main as _main

     sys.exit(_main())
diff --git a/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.py b/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.py
index 0ba06c52..6fdb59b7 100644
--- a/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.py
+++ b/pipenv/patched/pip/_internal/resolution/resolvelib/candidates.py
@@ -2,6 +2,7 @@ import logging
 import sys
 from typing import TYPE_CHECKING, Any, FrozenSet, Iterable, Optional, Tuple, Union, cast

+from pipenv.patched.pip._vendor.packaging.specifiers import SpecifierSet
 from pipenv.patched.pip._vendor.packaging.utils import NormalizedName, canonicalize_name
 from pipenv.patched.pip._vendor.packaging.version import Version

@@ -253,7 +253,10 @@ class _InstallRequirementBackedCandidate(Candidate):
        yield self._factory.make_requires_python_requirement(self.dist.requires_python)

     def get_install_requirement(self) -> Optional[InstallRequirement]:
-        return self._ireq
+        ireq = self._ireq
+        if self._version and ireq.req and not ireq.req.url:
+            ireq.req.specifier = SpecifierSet(f"=={self._version}")
+        return ireq


 class LinkCandidate(_InstallRequirementBackedCandidate):
diff --git a/pipenv/patched/pip/_internal/locations/__init__.py b/pipenv/patched/pip/_internal/locations/__init__.py
index 23eaea64..fe5dd5b4 100644
--- a/pipenv/patched/pip/_internal/locations/__init__.py
+++ b/pipenv/patched/pip/_internal/locations/__init__.py
@@ -41,7 +41,7 @@ logger = logging.getLogger(__name__)

 _PLATLIBDIR: str = getattr(sys, "platlibdir", "lib")

-_USE_SYSCONFIG_DEFAULT = sys.version_info >= (3, 10)
+_USE_SYSCONFIG_DEFAULT = sys.version_info >= (3, 7)


 def _should_use_sysconfig() -> bool:
