diff --git a/pipenv/patched/crayons.py b/pipenv/patched/crayons.py
index 455d3e9..7c9ff0d 100644
--- a/pipenv/patched/crayons.py
+++ b/pipenv/patched/crayons.py
@@ -12,14 +12,14 @@ import os
 import re
 import sys
 
-PY3 = sys.version_info[0] >= 3
-
 import colorama
 
+PY3 = sys.version_info[0] >= 3
+
 __all__ = (
     'red', 'green', 'yellow', 'blue',
-    'black', 'magenta', 'cyan', 'white',
-    'clean', 'disable'
+    'black', 'magenta', 'cyan', 'white', 'normal',
+    'clean', 'disable',
 )
 
 COLORS = __all__[:-2]
@@ -46,7 +46,7 @@ class ColoredString(object):
         self.color = color
         self.always_color = always_color
         self.bold = bold
-        if os.environ.get('CLINT_FORCE_COLOR'):
+        if os.environ.get('PIPENV_FORCE_COLOR'):
             self.always_color = True
 
     def __getattr__(self, att):
@@ -76,7 +76,6 @@ class ColoredString(object):
         else:
             return self.s
 
-
     def __len__(self):
         return len(self.s)
 
@@ -121,30 +120,43 @@ def clean(s):
     return txt
 
 
+def normal(string, always=False, bold=False):
+    return ColoredString('RESET', string, always_color=always, bold=bold)
+
+
 def black(string, always=False, bold=False):
     return ColoredString('BLACK', string, always_color=always, bold=bold)
 
+
 def red(string, always=False, bold=False):
     return ColoredString('RED', string, always_color=always, bold=bold)
 
+
 def green(string, always=False, bold=False):
     return ColoredString('GREEN', string, always_color=always, bold=bold)
 
+
 def yellow(string, always=False, bold=False):
     return ColoredString('YELLOW', string, always_color=always, bold=bold)
 
+
 def blue(string, always=False, bold=False):
     return ColoredString('BLUE', string, always_color=always, bold=bold)
 
+
 def magenta(string, always=False, bold=False):
     return ColoredString('MAGENTA', string, always_color=always, bold=bold)
 
+
 def cyan(string, always=False, bold=False):
     return ColoredString('CYAN', string, always_color=always, bold=bold)
 
+
 def white(string, always=False, bold=False):
+    # This upsets people...
     return ColoredString('WHITE', string, always_color=always, bold=bold)
 
+
 def disable():
     """Disables colors."""
     global DISABLE_COLOR
