diff --git a/pipenv/patched/pew/pew.py b/pipenv/patched/pew/pew.py
index 2d3889a0..07dbdf4f 100644
--- a/pipenv/patched/pew/pew.py
+++ b/pipenv/patched/pew/pew.py
@@ -215,6 +215,8 @@ def shell(env, cwd=None):
             inve(env, shell, '-c', shell_check)
         except CalledProcessError:
             return
+    if shell_name in ('Cmder', 'cmd'):
+        os.environ['PROMPT'] = '({0}) {1}'.format(env, os.environ['PROMPT'])
     if shell_name == 'bash':
         fork_bash(env, cwd)
     elif shell_name == 'Cmder':
