diff --git a/pipenv/patched/pip/_internal/index/collector.py b/pipenv/patched/pip/_internal/index/collector.py
index 00d66daa3..3cc20b1aa 100644
--- a/pipenv/patched/pip/_internal/index/collector.py
+++ b/pipenv/patched/pip/_internal/index/collector.py
@@ -394,9 +394,11 @@ class LinkCollector:
         self,
         session: PipSession,
         search_scope: SearchScope,
+        index_lookup: dict[str, list[str]] | None = None,
     ) -> None:
         self.search_scope = search_scope
         self.session = session
+        self.index_lookup = index_lookup if index_lookup else {}

     @classmethod
     def create(
@@ -404,6 +406,7 @@ class LinkCollector:
         session: PipSession,
         options: Values,
         suppress_no_index: bool = False,
+        index_lookup: dict[str, list[str]] | None = None,
     ) -> LinkCollector:
         """
         :param session: The Session to use to make requests.
@@ -425,10 +428,12 @@ class LinkCollector:
             find_links=find_links,
             index_urls=index_urls,
             no_index=options.no_index,
+            index_lookup=index_lookup,
         )
         link_collector = LinkCollector(
             session=session,
             search_scope=search_scope,
+            index_lookup=index_lookup,
         )
         return link_collector

diff --git a/pipenv/patched/pip/_internal/models/search_scope.py b/pipenv/patched/pip/_internal/models/search_scope.py
index 136163ca0..0101dd386 100644
--- a/pipenv/patched/pip/_internal/models/search_scope.py
+++ b/pipenv/patched/pip/_internal/models/search_scope.py
@@ -14,17 +14,17 @@ from pip._internal.utils.misc import normalize_path, redact_auth_from_url
 logger = logging.getLogger(__name__)


-@dataclass(frozen=True)
+@dataclass(frozen=False)
 class SearchScope:
     """
     Encapsulates the locations that pip is configured to search.
     """

-    __slots__ = ["find_links", "index_urls", "no_index"]
-
     find_links: list[str]
     index_urls: list[str]
     no_index: bool
+    index_lookup: dict[str, str] | None = None
+    index_restricted: bool | None = None

     @classmethod
     def create(
@@ -32,6 +32,8 @@ class SearchScope:
         find_links: list[str],
         index_urls: list[str],
         no_index: bool,
+        index_lookup: dict[str, list[str]] | None = None,
+        index_restricted: bool = False,
     ) -> "SearchScope":
         """
         Create a SearchScope object after normalizing the `find_links`.
@@ -66,6 +68,8 @@ class SearchScope:
             find_links=built_find_links,
             index_urls=index_urls,
             no_index=no_index,
+            index_lookup=index_lookup or {},
+            index_restricted=index_restricted,
         )

     def get_formatted_locations(self) -> str:
@@ -123,4 +127,9 @@ class SearchScope:
                 loc = loc + "/"
             return loc

-        return [mkurl_pypi_url(url) for url in self.index_urls]
+        index_urls = self.index_urls
+        if project_name in self.index_lookup:
+            index_urls = [self.index_lookup[project_name]]
+        elif self.index_restricted and self.index_urls:
+            index_urls = [self.index_urls[0]]
+        return [mkurl_pypi_url(url) for url in index_urls]
