diff --git a/pipenv/patched/prettytoml/elements/abstracttable.py b/pipenv/patched/prettytoml/elements/abstracttable.py
index 59fd5748..48663aed 100644
--- a/pipenv/patched/prettytoml/elements/abstracttable.py
+++ b/pipenv/patched/prettytoml/elements/abstracttable.py
@@ -1,8 +1,13 @@
+try:
+    from collections.abc import Mapping
+except ImportError:
+    from collections import Mapping
+
 from prettytoml.elements.common import ContainerElement
 from prettytoml.elements import traversal


-class AbstractTable(ContainerElement, traversal.TraversalMixin):
+class AbstractTable(ContainerElement, traversal.TraversalMixin, Mapping):
     """
     Common code for handling tables as key-value pairs with metadata elements sprinkled all over.

@@ -37,6 +42,9 @@ class AbstractTable(ContainerElement, traversal.TraversalMixin):
     def __len__(self):
         return len(tuple(self._enumerate_items()))

+    def __iter__(self):
+        return (key for key, _ in self.items())
+
     def __contains__(self, item):
         return item in self.keys()
